### PR TITLE
Implement select distinct

### LINKED ISSUE #250: Incorrect answers with SELECT DISTINCT queries
**Describe the bug**
`SELECT DISTINCT` (note not distinct aggregates, eg `SELECT COUNT (DISTINCT)...`) produce incorrect results

**To Reproduce**

```shell
echo "A" > /tmp/foo.csv
echo "B" >> /tmp/foo.csv
echo "B" >> /tmp/foo.csv
echo "B" >> /tmp/foo.csv
```

Then in the datafusion-cli:

```
> CREATE EXTERNAL TABLE t(col varchar)
STORED AS CSV
LOCATION '/tmp/foo.csv';
0 rows in set. Query took 0 seconds.
> SELECT DISTINCT col from t
> ;
+-----+
| col |
+-----+
| A   |
| B   |
| B   |
| B   |
+-----+
```

**Expected behavior**
Expected results are no duplicates. The output should contain only 2 rows with values of `A` and `B`

**Additional context**
I suspect we are just ignoring the `distinct` on these queries. We could "fix" the bug by generating an error about the feature being unsupported. Bonus points would be to add a group by operator on top of the plan to remove duplicates (e.g. group by all the items in the select list)

### PR DESCRIPTION
**Pull request by Dandandan**
# Which issue does this PR close?

Closes #250

 # Rationale for this change
Fix behavior and a simple implementation of select distinct.

# What changes are included in this PR?

An implementation of select distinct.

# Are there any user-facing changes?

A working select distinct

### CODE REVIEW

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -649,20 +657,15 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
     /// necessary, i.e., when the input fields are different than the
     /// projection. Note that if the input fields are the same, but out of
     /// order, the projection will be applied.
-    fn project(
-        &self,
-        input: &LogicalPlan,
-        expr: Vec<Expr>,
-        force: bool,
```
Comment (Dandandan) [inline datafusion/src/sql/planner.rs:656]: small drive-by simplification, an unused option

File: datafusion/tests/sql.rs [hunk 2]
```diff
@@ -372,6 +372,81 @@ async fn csv_query_group_by_float32() -> Result<()> {
...
+    let sql_all = "SELECT ALL c1 FROM aggregate_simple order by c1";
+    let actual_all = execute(&mut ctx, sql_all).await;
+
+    assert_eq!(actual_no_all, actual_all);
+
+    Ok(())
+}
+
+#[tokio::test]
+async fn select_distinct_all() -> Result<()> {
```
Comment (alamb) [inline datafusion/tests/sql.rs:392]:
```suggestion
async fn select_distinct() -> Result<()> {
```

File: datafusion/tests/sql.rs [hunk 3]
```diff
@@ -372,6 +372,81 @@ async fn csv_query_group_by_float32() -> Result<()> {
...
+
+#[tokio::test]
+async fn select_distinct_simple() -> Result<()> {
+    let mut ctx = ExecutionContext::new();
+    register_aggregate_simple_csv(&mut ctx)?;
+
+    let sql = "SELECT DISTINCT c1 FROM aggregate_simple order by c1";
+    let actual = execute(&mut ctx, sql).await;
+
+    let expected = vec![
```
Comment (alamb) [inline datafusion/tests/sql.rs:416]: I doubled checked that the input has duplicates (to show that the distinct has removed them):  https://github.com/apache/arrow-datafusion/blob/master/datafusion/tests/aggregate_simple.csv

👍

**Review by alamb — APPROVED** [review]
This looks great @Dandandan  thank you ❤️

**Review by andygrove — APPROVED** [review]
:tada:
